src/astar.rs

```diff
@@ -0,0 +1,168 @@
+/// let e = g.add_node((3., 3.));
+/// let f = g.add_node((4., 2.));
+/// let ab = g.add_edge(a, b, 2);
+/// let ad = g.add_edge(a, d, 4);
+/// let bc = g.add_edge(b, c, 1);
+/// let bf = g.add_edge(b, f, 7);
+/// let ce = g.add_edge(c, e, 5);
+/// let ef = g.add_edge(e, f, 1);
+/// let de = g.add_edge(d, e, 1);
+///
+/// let path = astar(&g, a, |finish| finish == f, |e| *e.weight(), |_| 0);
+/// assert_eq!(path, Some(vec![a, d, e, f]));
+/// ```
+///
+/// The graph should be `Visitable` and implement `IntoEdges`. Edge costs must be non-negative.
+///
+/// Returns a Path of subsequent `NodeId` from start to finish, if one was found.
+pub fn astar<G, F, H, K, IsGoal>(graph: G, start: G::NodeId, is_goal: IsGoal,
+                                     mut edge_cost: F, mut estimate_cost: H)
+    -> Option<Path<G>>
+    where G: IntoEdges + Visitable,
+          IsGoal: Fn(G::NodeId) -> bool,
+          G::NodeId: Eq + Hash,
+          F: FnMut(G::EdgeRef) -> K,
+          H: FnMut(G::NodeId) -> K,
+          K: Measure + Copy,
+{
+    let mut visited = graph.visit_map();
+    let mut visit_next = BinaryHeap::new();
+    let mut scores = HashMap::new();
+    let mut path_tracker = PathTracker::<G>::new();
+
+    let zero_score = K::default();
+    scores.insert(start, zero_score);
+    visit_next.push(MinScored(estimate_cost(start), start));
+
+    while let Some(MinScored(_, node)) = visit_next.pop() {
+        if is_goal(node) {
+            return Some(path_tracker.reconstruct_path_to(node));
+        }
+
+        // Don't visit the same node several times, as the first time it was visited it was using
+        // the shortest available path.
+        if !visited.visit(node) {
+            continue
+        }
+
+        // This lookup can be unwrapped without fear of panic since the node was necessarily scored
+        // before adding him to `visit_next`.
+        let node_score = *scores.get(&node).unwrap();
+
+        for edge in graph.edges(node) {
+            let next = edge.target();
+            if visited.is_visited(&next) {
+                continue
+            }
+
+            let mut next_score = node_score + edge_cost(edge);
+
+            match scores.entry(next) {
+                Occupied(ent) => {
+                    let old_score = *ent.get();
+                    if next_score < old_score {
+                        *ent.into_mut() = next_score;
+                        path_tracker.set_predecessor(next, node);
+                    } else {
+                        next_score = old_score;
+                    }
+                },
+                Vacant(ent) => {
+                    ent.insert(next_score);
+                    path_tracker.set_predecessor(next, node);
+                }
+            }
+
+            let next_estimate_score = next_score + estimate_cost(next);
+            visit_next.push(MinScored(next_estimate_score, next));
+        }
+    }
+
+    None
+}
+
+pub type Path<G: GraphBase> = Vec<G::NodeId>;
```

Comment:
probably rewrite this without the trait bound and with `<G as GraphBase>::NodeId` instead. That's the form the compiler doesn't warn about right now at least..

Reply:
Possibly this should just be a private type alias. Then it's just a Vec in the public docs.

Reply:
I removed the `Path` alias for now, since it's just a `Vec`.